### PR TITLE
added a tool for computing a lineage table like gi2taxonomy (but better)

### PR DESCRIPTION
**Pull request by bernt-matthias**
The python script:
- is ete3 based
- outputs either all levels or just the main levels
- if main levels are undefined in a linage then the info can be taken from sub, super, ... taxons

some points for the discussion:
- most important: both python scripts now use an sqlite data base as parameter (instead of a per user DB in ~/.etetoolkit/ which is created by ete3. When multiple users use galaxy an update can be triggered by one user which would affect the reproducibility for other users.)
- One would now needs an additional tool to create the ETE sqlite DB from the NCBI taxonomy DB. Any suggestions? An extra Data Source tool or extend the NCBI taxonomy Data Source.
- the name2taxid lookup takes only the first returned result. if names can be resolved (eg Actinobacteria) this is confusing. one could add an option that allows to use the lower/higher taxon. maybe it would be better to create an extra tool to create an name -> taxid mapping.
- "root" is excluded (which is different from gi2taxonomy) 
- fuzzy name search lookup could be enabled, but this would require a special build of pysqlite  where extensions are supported (which is not the default)

### CODE REVIEW

**Comment by bernt-matthias** [comment]
TODOs: 
- I could not add the taxa.sqlite because it is to large for github. Is there a solution? 
- Currently the test for the tree generator fails and I do not understand why. 

What do you think: @nsoranzo @anilthanki

**Comment by bernt-matthias** [comment]
Note the ete version where a sqlite DB creation from the NCBI dump is possible is apparently not available via pip: https://github.com/etetoolkit/ete/issues/282

File: tools/ete/ete_lineage_generator.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    """
+    lineage = ncbi.get_lineage(taxid)
+    lineage_ranks = ncbi.get_rank(lineage)
+    lineage_names = ncbi.get_taxid_translator(lineage, try_synonyms=True)
+    if lower:
+        lineage.reverse()
+    for l in lineage:
+        if not lineage_ranks[l] in RANK_IDX:
+            continue
+        if ranks[ RANK_IDX[lineage_ranks[l]]] != "NA":
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.py]: Can you remove the space between `[` and `RANK_IDX` to make flake8 happy, please?

**Comment by nsoranzo** [comment]
> I could not add the taxa.sqlite because it is to large for github. Is there a solution?

@bernt-matthias Can you reduce the size of the database by deleting the table/rows that are not used in the tests?

File: tools/ete/ete_init_taxdb.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import sys
+sys.path.append("./ete/")
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: Why is that needed?

File: tools/ete/ete_init_taxdb.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+
+import ete3
+
+ete3.NCBITaxa(taxdump_file="/tmp/taxdump/taxdump.tar.gz", dbfile="taxdump.sqlite")
+
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: Remove blank line for flake8.

**Comment by bernt-matthias** [comment]
The testdata contains only the lineages of the input data.

Note that the sqlitedb can be generated with ete_init_taxdb.py only by using my fork of ete3 (https://github.com/bernt-matthias/ete PR is pending for quite a while). One could make also a hack using the normal ete3 library by calling NCBITaxa function only with the tar.gz file and then moving the sqlite file from ~/.etetoolkit ...

File: tools/ete/ete_init_taxdb.py (hunk 2 above)
Comment (bernt-matthias) [reply tools/ete/ete_init_taxdb.py]: see my comment below (the official ete code does not work like this)

File: tools/ete/ete_init_taxdb.py (hunk 3 above)
Comment (bernt-matthias) [reply tools/ete/ete_init_taxdb.py]: done

**Comment by bernt-matthias** [comment]
the remaining question is how we integrate the sqlite db in the tool? data manager?

File: tools/ete/ete_lineage_generator.xml [hunk 4]
```diff
@@ -0,0 +1,143 @@
+<tool id="ete_linage_generator" name="ETE lineage generator" version="3.0.0b35">
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: s/linage/lineage/

**Comment by nsoranzo** [comment]
> the remaining question is how we integrate the sqlite db in the tool? data manager?

Do you want the user to be able to select between multiple versions of the database?

> Note that the sqlitedb can be generated with ete_init_taxdb.py only by using my fork of ete3 (https://github.com/bernt-matthias/ete PR is pending for quite a while).

Do you mean https://github.com/etetoolkit/ete/pull/288 ? I can't see why is that related. Or do you mean release 3.0.0b36 , which is not available on PyPI yet as you reported in https://github.com/etetoolkit/ete/issues/282 ?

File: tools/ete/ete_lineage_generator.xml [hunk 5]
```diff
@@ -0,0 +1,143 @@
+<tool id="ete_lineage_generator" name="ETE lineage generator" version="3.0.0b35">
+		<description>from a list of species/taxids using the ETE Toolkit</description>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Indentation. Please always use 4-spaces instead of tabs.

File: tools/ete/ete_lineage_generator.xml [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+     <requirements>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Use 4-spaces instead of 5

File: tools/ete/ete_lineage_generator.xml [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        <requirement type="package" version="3.0.0b35">ete3</requirement>
+    </requirements>
+    <stdio>
+        <!-- Anything other than zero is an error -->
+        <exit_code range="1:" />
+    </stdio>
+    <command>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: You can add `detect_errors="exit_code"` attribute to `<command>` and remove the whole `<stdio>` block.

File: tools/ete/ete_lineage_generator.xml [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    <![CDATA[
+python $__tool_directory__/ete_lineage_generator.py
+-s '$speciesFile'
+-d '$database'
+-o '$outputFile'
+#if $ranks.levels == "full"
+-f
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Can you add (4-spaces) indentation also to the Cheetah code, please?

**Comment by bernt-matthias** [comment]
> Do you want the user to be able to select between multiple versions of the database?

Yes, I think that this would be a good idea. (In the current tool the sqlite DB in ~/.etetoolkit is used which is particularly problematic for when jobs are started as real user)

**Comment by bernt-matthias** [comment]
> Do you mean etetoolkit/ete#288 ? I can't see why is that related. Or do you mean release 3.0.0b36 , which is not available on PyPI yet as you reported in etetoolkit/ete#282 ?

I think I mixed up things. Indeed I mean https://github.com/etetoolkit/ete/issues/282

File: tools/ete/ete_lineage_generator.xml [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+			        <option value="species subgroup">species subgroup</option>
+					<option value="species" selected="True">species</option>
+					<option value="subspecies">subspecies</option>
+                    <option value="varietas">varietas</option>
+			        <option value="forma">forma</option>
+	            </param>
+            </when>
+	</conditional>
+		<conditional name="compresscond">
+				<param name="compress" type="boolean" label="Fill unnamed ranks" truevalue="-c" falsevalue="" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Please use a `type="select"` for the conditional param, see https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#booleans .

File: tools/ete/ete_lineage_generator.xml [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+				<when value="-c">
+					<param name="lower" type="boolean" label="Prefer lower ranks for filling" truevalue="-l" falsevalue="" help="Prefer lower levels for filling" />
+				</when>
+		</conditional>
+    </inputs>
+    <outputs>
+        <data format="tsv" name="outputFile" label="${tool.name} on ${on_string}"/>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Move `format` after `name`.

File: tools/ete/ete_lineage_generator.xml [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+					<option value="subspecies">subspecies</option>
+                    <option value="varietas">varietas</option>
+			        <option value="forma">forma</option>
+	            </param>
+            </when>
+	</conditional>
+		<conditional name="compresscond">
+				<param name="compress" type="boolean" label="Fill unnamed ranks" truevalue="-c" falsevalue="" help="Fill unnamed ranks with super/sub ranks (see -l)" />
+				<when value="-c">
+					<param name="lower" type="boolean" label="Prefer lower ranks for filling" truevalue="-l" falsevalue="" help="Prefer lower levels for filling" />
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: `label` and `help` should always be the second-to-last and last attributes respectively, according to IUC [coding style](https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html#coding-style). Rationale is that if the XML element is long, `label` and `help` attributes can be on a new line(s).

**Comment by nsoranzo** [comment]
> > Do you want the user to be able to select between multiple versions of the database?

> Yes, I think that this would be a good idea. (In the current tool the sqlite DB in ~/.etetoolkit is used which is particularly problematic for when jobs are started as real user)

So a good start would be to add a data table (adding a `tool_data_table_conf.xml.sample` and a .`loc` file in `tool-data/` ) to be used for the `database` param with the `<options from_data_table="...">` syntax.

See https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/integration_checklist.html#creating-the-tool-wrapper

**Comment by bernt-matthias** [comment]
I decided against the tool_data_table_conf option. One can still get the data base as library. 

I added a tool for creating the sqlite DB. But this wont work until https://github.com/etetoolkit/ete/issues/282 is fixed. 

There are currently lots of errors popping up in planemo test and serve: 
```
2017-09-12 11:21:30,000 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has not been defined for 'ranks (levels) --> full', assuming empty inputs.
2017-09-12 11:21:30,000 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has not been defined for 'ranks (levels) --> primary', assuming empty inputs.
2017-09-12 11:21:30,001 WARNI [galaxy.tools] Tool ete_lineage_generator: a when tag has been defined for 'compresscond (compress) --> -c', but does not appear to be selectable.
```

`ValueError: ('No case matched value:', 'compresscond', 'None')
`

Any suggestion what I did wrong?

File: tools/ete/ete_init_taxdb.xml [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    </command>
+    <inputs>
+        <conditional name="latestcond">
+            <param name="latest" type="select" label="Get latest version" help="Get the latest version of the taxdump from the NCBI website">
+                <option value="local" selected="True">Use a local taxdump</option>
+                <option value="latest">Get latest taxdump online</option>
+            </param>
+            <when value="local">
+                <param name="taxdump" type="data" format="txt" label="NCBI taxdump" help="NCBI taxonomy database dump in tar.gz format" />
+            </when>
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: Add:
```xml
            <when value="latest" />
```

File: tools/ete/ete_lineage_generator.xml [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+-->
+    <inputs>
+        <param name="speciesFile" type="data" format="txt" label="Species file" help="List with one species/taxid per line" />
+        <param name="database" type="data" format="sqlite" label="(ETE3) Taxonomy Database" help="The sqlite formatted Taxonomy used by ETE3 (which is derived from NCBI taxonomy)" />
+        <conditional name="ranks">
+            <param name="levels" type="select" label="Taxonomic levels" help="Taxonomic levels to include in the output">
+                <option value="primary" selected="true">Primary</option>
+                <option value="full">Full</option>
+                <option value="manual">Manual</option>
+            </param>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Add:
```xml
            <when value="primary" />
            <when value="full" />
```

**Comment by bernt-matthias** [comment]
And one more question: how to handle the tar.gz file correctly? Is "text" OK?

File: tools/ete/ete_lineage_generator.xml [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+                    <option value="species subgroup">species subgroup</option>
+                    <option value="species" selected="True">species</option>
+                    <option value="subspecies">subspecies</option>
+                    <option value="varietas">varietas</option>
+                    <option value="forma">forma</option>
+                </param>
+            </when>
+        </conditional>
+        <conditional name="compresscond">
+                <param name="compress" type="select" truevalue="-c" falsevalue="" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: That's an hybrid now ;)
It should be:
```xml
            <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)">
                <option value="-c">Yes</option>
                <option value="">No</option>
            </param>
```

File: tools/ete/ete_lineage_generator.xml [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+                <when value="-c">
+                    <param name="lower" type="select" label="Prefer lower ranks for filling" help="Prefer lower levels for filling">
+                        <option value="-l">Yes</option>
+                        <option value="">No</option>
+                    </param>
+                </when>
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Add:
```xml
                <when value="" />
```

File: tools/ete/ete_species_tree_generator.xml [hunk 16]
```diff
@@ -11,17 +11,19 @@
...
 #else
     -f ${output_format.format_selector}
 #end if
 -t $output_format.treebest
--d $database
     ]]>
     </command>
     <inputs>
         <param name="speciesFile" type="data" format="txt" label="Species file" help="List with one species per line" />        
+		<param name="database" type="data" format="sqlite" label="(ETE3) Taxonomy Database" help="The sqlite formatted Taxonomy used by ETE3 (which is derived from NCBI taxonomy)" />
```
Comment (nsoranzo) [inline tools/ete/ete_species_tree_generator.xml]: Don't use tabs, please!

**Comment by bgruening** [comment]
You could use: https://github.com/galaxyproject/galaxy/blob/dev/config/datatypes_conf.xml.sample#L200

**Comment by nsoranzo** [comment]
> I decided against the tool_data_table_conf option. One can still get the data base as library.
> I added a tool for creating the sqlite DB.

That's fine for me, I actually wanted to suggest you something like that, but forgot yesterday.

I had review comments that should fix the errors you're seeing.

File: tools/ete/ete_lineage_generator.xml [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+                    <option value="subspecies">subspecies</option>
+                    <option value="varietas">varietas</option>
+                    <option value="forma">forma</option>
+                </param>
+            </when>
+            <when value="primary" />
+            <when value="full" />
+        </conditional>
+        <conditional name="compresscond">
+                <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)" />
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Still not right, this line should be substituted with:
```xml
            <param name="compress" type="select" label="Fill unnamed ranks" help="Fill unnamed ranks with super/sub ranks (see -l)">
                <option value="-c">Yes</option>
                <option value="">No</option>
            </param>
```

Also the indentation inside this conditional block is wrong.

File: tools/ete/ete_init_taxdb.py [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+parser = optparse.OptionParser()
+parser.add_option('-t', '--taxdump', dest='taxdump', default=None,
+                  help='NCBI taxdump (tar.gz)')
+parser.add_option('-d', '--database', dest="database", default=None,
+                  help='ETE sqlite data base to use (default: ~/.etetoolkit/taxa.sqlite)')
+options, args = parser.parse_args()
+if options.database is None:
+    parser.error("-d option must be specified")
+if options.taxdump is not None:
+    taxdump = options.database
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: `options.taxdump` should go here, I suppose.

File: tools/ete/ete_init_taxdb.xml [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+            <when value="latest" />
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="database" format="sqlite" label="${tool.name} on ${on_string}" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="database" ftype="sqlite" value="taxdump.sqlite" />
+            <output name="taxdump" file="taxdump.tar.gz" />
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: You seem to have swapped the param and output here.

File: tools/ete/ete_init_taxdb.xml [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+            </when>
+            <when value="latest" />
+        </conditional>
+    </inputs>
+    <outputs>
+        <data name="database" format="sqlite" label="${tool.name} on ${on_string}" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="taxdump" ftype="tar" file="taxdump.tar.gz" />
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: s/file/value/

File: tools/ete/ete_init_taxdb.xml [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+            <output name="database" value="taxdump.sqlite" />
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: s/value/file/

**Comment by nsoranzo** [comment]
Tests are failing because Planemo cannot validate 2 URLs, will be fixed by https://github.com/galaxyproject/planemo/pull/725 .

**Comment by bernt-matthias** [comment]
I guess its nearly done. Still have some problems with the tests.

File: tools/ete/ete_init_taxdb.py [hunk 22]
```diff
@@ -0,0 +1,26 @@
+import optparse
+import sys
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: This import is unused, please remove it (or comment it out as the next line).

File: tools/ete/ete_init_taxdb.py [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+if options.database is None:
+    parser.error("-d option must be specified")
+if options.taxdump is not None:
+    taxdump = options.taxdump
+else:
+    from urllib import urlretrieve
+    urlretrieve("http://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz", "taxdump.tar.gz")
+    taxdump = "taxdump.tar.gz"
+
+# will remove a taxdump.tar.gz file at the end
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: Remove trailing whitespace.

File: tools/ete/ete_init_taxdb.py [hunk 24]
```diff
@@ -0,0 +1,26 @@
...
+if options.taxdump is not None:
+    taxdump = options.taxdump
+else:
+    from urllib import urlretrieve
+    urlretrieve("http://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz", "taxdump.tar.gz")
+    taxdump = "taxdump.tar.gz"
+
+# will remove a taxdump.tar.gz file at the end 
+# which will lead to an errmessage if not present
+# if the tool is run on a taxdump in the current dir it will be
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: Remove trailing whitespace.

**Review by nsoranzo — COMMENTED** [review]
3 linting fixes needed, then we'll see the proper tool tests

File: tools/ete/ete_lineage_generator.xml [hunk 25]
```diff
@@ -0,0 +1,153 @@
...
+        <test>
+            <param name="speciesFile" ftype="txt" value="species.txt" />
+            <param name="database" ftype="sqlite" value="taxdump.sqlite" />
+            <param name="compress" value="" />
+            <param name="levels" value="primary" />
+            <output name="outputFile" file="lineage.txt" />
+        </test>
+        <test>
+            <param name="speciesFile" ftype="txt" value="species.txt" />
+            <param name="database" ftype="sqlite" value="taxsump.sqlite" />
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: s/taxsump/taxdump/

File: tools/ete/ete_init_taxdb.xml [hunk 26]
```diff
@@ -0,0 +1,54 @@
+<tool id="ete_init_taxdb" name="ETE taxa DB generator" version="3.0.0b35">
+    <description>generates the ETE sqlite DB from the NCBI taxdump.tar.gz</description>
+    <requirements>
+        <requirement type="package" version="3.0.0b35">ete3</requirement>
+    </requirements>
+    <command detect_errors="exit_code">
+    <![CDATA[
+python $__tool_directory__/ete_init_taxdb.py
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: Please add single quotes around `$__tool_directory__/ete_init_taxdb.py`

File: tools/ete/ete_init_taxdb.xml [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+#if $latest == "local"
+    -t '$taxdump'
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.xml]: s/$taxdump/$latestcond.taxdump/

File: tools/ete/ete_init_taxdb.py [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+                  help='NCBI taxdump (tar.gz) will be downloaded if not given')
+parser.add_option('-d', '--database', dest="database", default=None,
+                  help='ETE sqlite data base to use (default: ~/.etetoolkit/taxa.sqlite)')
+options, args = parser.parse_args()
+if options.database is None:
+    parser.error("-d option must be specified")
+if options.taxdump is not None:
+    taxdump = options.taxdump
+else:
+    from urllib import urlretrieve
```
Comment (nsoranzo) [inline tools/ete/ete_init_taxdb.py]: For Python3 compatibility, can you please change this to:
```
from six.moves.urllib.request import urlretrieve
```
and move it to the top, just after the `ete3.ncbi_taxonomy` import?

File: tools/ete/ete_lineage_generator.xml [hunk 29]
```diff
@@ -0,0 +1,153 @@
+<tool id="ete_lineage_generator" name="ETE lineage generator" version="3.0.0b35">
+    <description>from a list of species/taxids using the ETE Toolkit</description>
+    <requirements>
+        <requirement type="package" version="3.0.0b35">ete3</requirement>
+    </requirements>
+    <command detect_errors="exit_code">
+    <![CDATA[
+python $__tool_directory__/ete_lineage_generator.py
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: Add single-quotes around `$__tool_directory__/ete_lineage_generator.py`

File: tools/ete/ete_lineage_generator.xml [hunk 30]
```diff
@@ -0,0 +1,153 @@
...
+-s '$speciesFile'
+-d '$database'
+-o '$outputFile'
+#if $levels == "full"
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: s/$levels/$ranks.levels/

File: tools/ete/ete_lineage_generator.xml [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+    -f 
+#elif $levels == "manual"
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: s/$levels/$ranks.levels/

File: tools/ete/ete_lineage_generator.xml [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+    #for $rank in $manualranks:
```
Comment (nsoranzo) [inline tools/ete/ete_lineage_generator.xml]: s/$manualranks/$ranks.manualranks/

**Review by nsoranzo — COMMENTED** [review]
Let me know if you need help with setting up [Planemo](https://planemo.readthedocs.io/), `planemo shed_lint` and `planemo test` can really speed up tool development!

**Comment by bernt-matthias** [comment]
Still struggling with planemo test. Somehow there is a difference for the database and the generated tree. I never used shed_lint. Where can I start to read?

Thanks for your help.. learning slowly.

File: tools/ete/ete_init_taxdb.py [hunk 33]
```diff
@@ -0,0 +1,23 @@
+import ete3.ncbi_taxonomy
+import optparse
```
Comment (anilthanki) [inline tools/ete/ete_init_taxdb.py]: import optparse should be before ete3.ncbi_taxonomy

Comment (nsoranzo) [reply tools/ete/ete_init_taxdb.py]: Yep, we (as `galaxy` and `tools-iuc`) use the "smarkets" style, which divides imports in 4 groups (separated by a single blank line):

- `from __future__ import ...`
- standard library imports
- third-party imports
- application-specific imports

Within each grouping, imports should be sorted lexicographically, ignoring case, according to each module's full package path, but with `import ...` statements before `from X import ...` statements, see an [example](https://github.com/PyCQA/flake8-import-order/blob/master/tests/test_cases/complete_smarkets.py).

So, in this case it should be:

```python
import optparse

import ete3.ncbi_taxonomy
from six.moves.urllib.request import urlretrieve
```

**Comment by nsoranzo** [comment]
Planemo documentation is at http://planemo.readthedocs.io/ , the commands we use are:
```
planemo shed_lint --tools --ensure_metadata --urls --report_level warn --fail_level error --recursive TOOL_DIR
planemo test --conda_dependency_resolution --conda_auto_install --galaxy_branch release_17.05 TOOL_DIR
```

**Comment by bernt-matthias** [comment]
Import order is fixed. `planemo shed_lint` is successful, but `planemo test` still has problems with ete_init and ete_species_tree_generator. 

For ete_species_tree_generator the comparison of `... test-data/out.nhx v. /tmp/tmpTglW3_out.nhx` is not successful because the file in /tmp seems to be absent. The actual output seems to be correct.

**Comment by nsoranzo** [comment]
> For ete_species_tree_generator the comparison of ... test-data/out.nhx v. /tmp/tmpTglW3_out.nhx is not successful because the file in /tmp seems to be absent. The actual output seems to be correct.

That's not your fault, the NCBI database has been updated and the tree is slightly different now. You can update the test output data with:
```
planemo test --update_test_data /PATH/TO/earlham-galaxytools/tools/ete/ete_species_tree_generator.xml
```
and then commit the changes to `tools/ete/test-data/out.nhx` . I can do that, but then you would have to rebase this PR.

**Comment by bernt-matthias** [comment]
The problem with the species tree generator seems to be fixed. 

The init DB test still fails despite updated test data. Since the test works on a local dump of (a part of the taxonomy) a change of the online NCBI taxonomy should be not the reason. I guess that using diff on the binary sqlite files could be the problem. Are there other possibilities to test. 

It would also be nice to have a test for the download. Can one just test for existence / file size larger X?

**Comment by nsoranzo** [comment]
@bernt-matthias Thanks for updating the `ete_species_tree_generator` test data.

Probably the best choice is add `compare="sim_size"` to the `<output />` element in the failing test. For more details, see https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output

**Comment by bernt-matthias** [comment]
OK - Got it. If I would like to add a test for the download of the complete taxdb how could I create a useful test without having the complete taxdump-sqlite in the test-data directory? 

The simplest test would be to test for the existence of the files. How would this be done? Alternatively there is a md5 at ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz.md5 that should match the taxdump.tar.gz file (but the taxdump is not the output, but the sqlite). Could this be integrated?

**Comment by nsoranzo** [comment]
@bernt-matthias The possible tests are the one in the link I gave above https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output , so what I have in my mind is probably not possible at the moment.
On the other hand, tests are supposed to be fast and with small inputs/outputs, so what we have seems good enough to me!

**Comment by bernt-matthias** [comment]
Wonderful. Thanks. 

Btw. Next week I will give a small workshop on Galaxy and wrapping galaxy tools. I will suggest there to wrap some of the ete3 command line tools...

**Comment by bernt-matthias** [comment]
I "finished" a wrapper for `ete mod`. There is still an error popping up due to the `--help` output in `<help>`. Could someone tell me what the problem and the solution is? I guess some quoting is necessary. Otherwise the tool seems be finished apart from a few lines of documentation in the help. 

I currently do not have the time to wrap `ete3 compare` and `ete3 view` which would be really nice.

File: tools/ete/ete_mod.xml [hunk 34]
```diff
@@ -0,0 +1,170 @@
...
+            <output name="output_tree" file="tree1_unroot_resolve.nhx" />
+        </test>
+        <test>
+            <param name="input_tree" ftype="nhx" value="tree3.nhx" />
+            <param name="standardize" value="--standardize" />
+            <output name="output_tree" file="tree3_stand.nhx" />
+        </test>
+    </tests>
+    <help><![CDATA[
+usage: ete3 mod [-h] [-t [SRC_TREES [SRC_TREES ...]]]
```
Comment (nsoranzo) [inline tools/ete/ete_mod.xml]: @bernt-matthias I don't think it's useful to have the tool help copied in the `<help>` section, it's better to have a general description of what the tool does and a link to the upstream documentation. Each option should be documented using the `help` attribute of the `<param>`s.

**Comment by bernt-matthias** [comment]
Agreed. Finished from my side.

File: tools/ete/ete_mod.xml [hunk 35]
```diff
@@ -0,0 +1,125 @@
...
+        <citation type="bibtex">
+@article{huerta2016ete,
+  title={ETE 3: reconstruction, analysis, and visualization of phylogenomic data},
+  author={Huerta-Cepas, Jaime and Serra, Fran{\c{c}}ois and Bork, Peer},
+  journal={Molecular Biology and Evolution},
+  volume={33},
+  number={6},
+  pages={1635-1638},
+  year={2016}
+}
```
Comment (nsoranzo) [inline tools/ete/ete_mod.xml]: You can replace with:
```
<citation type="doi">10.1093/molbev/msw046</citation>
```

Actually it starts to make sense to have a macro file for requirements and citations.

**Comment by nsoranzo** [comment]
@bernt-matthias I've rebased the pull request on top of latest master to include 2 fixes for `ete_species_tree_generator`. I've also made some cleanup and added macros for version, requirements and citations.
I hope that's OK with you. I'm going to (finally!) merge this now.

Thanks a lot for this contribution!
